### PR TITLE
Make Triage local clustering multithreaded

### PR DESCRIPTION
**Pull request by michaelkolber**
ref: #18672

/cc BenTheElder spiffxp

We'll want to compare both timing difference and accuracy difference. I've saved the current version of the results (as they get overwritten every time Triage runs within a day).

### CODE REVIEW

File: triage/summarize/cluster.go [hunk 1]
```diff
@@ -60,17 +67,81 @@ func clusterLocal(failuresByTest failuresGroup) nestedFailuresGroups {
...
+		Each worker will pull from the work queue and perform a clustering operation, and then push
+		the result to the done queue.
+	*/
+
+	var workerWG sync.WaitGroup
+	var doneWG sync.WaitGroup // Prevents this function from returning results before they are all written to clustered
+
+	numWorkers := threads
+	if threads <= 0 {
+		numWorkers = runtime.NumCPU() - 1 // Dedicate one CPU to pushing to the work queue/pulling from the done queue/main thread
```
Comment (BenTheElder) [inline triage/summarize/cluster.go:89]: suggestion: I usually recommend hauling references to globals up the stack, it makes it easier to unit test etc. later

I would put this as the default for a flag that allows tuning this.

also: what if NumCPU is 1?

Comment (michaelkolber) [reply triage/summarize/cluster.go:89]: I set the flag default as max of NumCPU-1 or 1, and then also set a (I think sensible) default of 1 for the function if numWorkers is passed in as <= 0.

**Comment by BenTheElder** [comment]
/lgtm
/approve
once this merges, the next step is to set CPU requests and limits on the job (matching to guarantee CPU) and set the flag to match

File: triage/summarize/cluster.go [hunk 2]
```diff
@@ -60,17 +66,79 @@ func clusterLocal(failuresByTest failuresGroup) nestedFailuresGroups {
...
+
+	// Read from the done queue
+	doneQueueWG.Add(1)
+	go func() {
+		defer doneQueueWG.Done()
+
+		for dg := range doneQueue {
+			numFailures += len(dg.input.Failures)
+			logInfo("%4d/%4d tests, %5d failures, %s", len(clustered)+1, len(failuresByTest), len(dg.input.Failures), dg.input.Key)
+			clustered[dg.input.Key] = dg.output
```
Comment (BenTheElder) [inline triage/summarize/cluster.go:117]: ah, I'm surprised this wasn't caught by the unit tests running with race detector enabled,

I missed that this is the same map, you can't have concurrent map read/write without a mutex.

Comment (BenTheElder) [reply triage/summarize/cluster.go:117]: Actually it's in clusterTest, so it's not obvious in the patchset.
https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-test-infra-triage/1294155497288503297

Comment (michaelkolber) [reply triage/summarize/cluster.go:117]: Ah, yeah, I should have followed the function calls all the way down.

Comment (michaelkolber) [reply triage/summarize/cluster.go:117]: Addressed in #18845

**Pull request merged**